"flutter doctor --android-licenses" doesn't gracefully handle old Android SDKs

```
ianh@ianh:~/dev$ flutter upgrade
Building flutter tool...
Upgrading Flutter from /usr/local/google/home/ianh/dev/flutter...
Already up-to-date.

Upgrading engine...
Already up-to-date.

Flutter • channel master • git@github.com:flutter/flutter.git
Framework • revision fc1b5c9ba9 (3 hours ago) • 2017-05-10 07:45:16 -0700
Engine • revision cbbf956fcb
Tools • Dart 1.23.0-dev.11.11

Running flutter doctor...
[✓] Flutter (on Linux, channel master)
    • Flutter at /usr/local/google/home/ianh/dev/flutter
    • Framework revision fc1b5c9ba9 (3 hours ago), 2017-05-10 07:45:16 -0700
    • Engine revision cbbf956fcb
    • Tools Dart version 1.23.0-dev.11.11

[✓] Android toolchain - develop for Android devices (Android SDK 25.0.2)
    • Android SDK at /usr/local/google/home/ianh/Android/Sdk
    • Platform android-25, build-tools 25.0.2
    • Java binary at: /usr/local/google/home/ianh/bin/android-studio/jre/bin/java
    • Java version: OpenJDK Runtime Environment (build 1.8.0_76-release-b03)

[✓] Android Studio (version 2.2)
    • Android Studio at /usr/local/google/home/ianh/bin/android-studio
    • Gradle version 2.14.1
    • Java version: OpenJDK Runtime Environment (build 1.8.0_76-release-b03)

[✓] Connected devices
    • Nexus 6 • ZX1G22KJ3C • android-arm • Android 7.1.1 (API 25)

ianh@ianh:~/dev$ flutter create dummy; cd dummy; flutter run
Creating project dummy...
Wrote 53 files.

Running 'flutter packages get' in dummy...            2.3s

[✓] Flutter is fully installed. (on Linux, channel master)
[✓] Android toolchain - develop for Android devices is fully installed. (Android SDK 25.0.2)
[✓] Android Studio is fully installed. (version 2.2)
[✓] Connected devices is fully installed.

All done! In order to run your application, type:

  $ cd dummy
  $ flutter run

Your main program file is lib/main.dart in the dummy directory.

Launching lib/main.dart on Nexus 6 in debug mode...
FAILURE: Build failed with an exception.

* Where:
Build file '/usr/local/google/home/ianh/dev/dummy/android/build.gradle' line: 20

* What went wrong:
A problem occurred evaluating root project 'android'.
> A problem occurred configuring project ':app'.
   > Failed to notify project evaluation listener.
      > You have not accepted the license agreements of the following SDK components:
        [Android SDK Build-Tools 25.0.3].
        Before building your project, you need to accept the license agreements and complete the installation of the missing components using the Android Studio SDK Manager.
        Alternatively, to learn how to transfer the license agreements from one workstation to another, go to http://d.android.com/r/studio-ui/export-licenses.html
      > Could not get unknown property 'compileDebugJavaWithJavac' for project ':app' of type org.gradle.api.Project.

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output.
Error running gradle: Exit code 1 from: /usr/local/google/home/ianh/bin/android-studio/gradle/gradle-2.14.1/bin/gradle app:properties
Running 'gradle assembleDebug'...                         
Checking the license for package Android SDK Build-Tools 25.0.3 in /usr/local/google/home/ianh/Android/Sdk/licenses
Warning: License for package Android SDK Build-Tools 25.0.3 not accepted.

FAILURE: Build failed with an exception.

* Where:
Build file '/usr/local/google/home/ianh/dev/dummy/android/build.gradle' line: 20

* What went wrong:
A problem occurred evaluating root project 'android'.
> A problem occurred configuring project ':app'.
   > Failed to notify project evaluation listener.
      > You have not accepted the license agreements of the following SDK components:
        [Android SDK Build-Tools 25.0.3].
        Before building your project, you need to accept the license agreements and complete the installation of the missing components using the Android Studio SDK Manager.
        Alternatively, to learn how to transfer the license agreements from one workstation to another, go to http://d.android.com/r/studio-ui/export-licenses.html
      > Could not get unknown property 'compileDebugJavaWithJavac' for project ':app' of type org.gradle.api.Project.

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output.
Gradle build failed: 1
ianh@ianh:~/dev/dummy$ flutter doctor --help
Show information about the installed tooling.

Usage: flutter doctor [arguments]
-h, --help                     Print this usage information.
    --[no-]android-licenses    Run the Android SDK manager tool to accept the SDK's licenses.

Run "flutter help" to see global options.
ianh@ianh:~/dev/dummy$ flutter doctor --android-licenses
Error: Unknown argument --licenses
Usage: 
  sdkmanager [--uninstall] [<common args>] \
    [--package_file <package-file>] [<packages>...]
  sdkmanager --update [<common args>]
  sdkmanager --list [<common args>]

In its first form, installs, or uninstalls, or updates packages.
    <package> is a sdk-style path (e.g. "build-tools;23.0.0" or 
             "platforms;android-23").
    <package-file> is a text file where each line is a sdk-style path
                   of a package to install or uninstall.
    Multiple --package_file arguments may be specified in combination
     with explicit paths.
In its second form (with --update), currently installed packages are
    updated to the latest version.
In its third form, all installed and available packages are printed out.

Common Arguments:
    --sdk_root=<sdkRootPath>: Use the specified SDK root instead of the SDK containing this tool
    --channel=<channelId>: Include packages in channels up to <channelId>.
                           Common channels are:
                           0 (Stable), 1 (Beta), 2 (Dev), and 3 (Canary).

    --include_obsolete: With --list, show obsolete packages in the
                        package listing. With --update, update obsolete
                        packages as well as non-obsolete.
    --no_https: Force all connections to use http rather than https.
    --proxy=<http | socks>: Connect via a proxy of the given type.
    --proxy_host=<IP or DNS address>: IP or DNS address of the proxy to use.
    --proxy_port=<port #>: Proxy port to connect to.

* If the env var REPO_OS_OVERRIDE is set to "windows",
  "macosx", or "linux", packages will be downloaded for that OS.
ianh@ianh:~/dev/dummy$ 

```

--- a/packages/flutter_tools/lib/src/android/android_sdk.dart
+++ b/packages/flutter_tools/lib/src/android/android_sdk.dart
@@ -347,7 +347,8 @@ class AndroidSdk {
       throwToolExit('Android sdkmanager not found. Update to the latest Android SDK to resolve this.');
     final ProcessResult result = processManager.runSync(<String>[sdkManagerPath, '--version'], environment: sdkManagerEnv);
     if (result.exitCode != 0) {
-      throwToolExit('sdkmanager --version failed: ${result.exitCode}', exitCode: result.exitCode);
+      printTrace('sdkmanager --version failed: exitCode: ${result.exitCode} stdout: ${result.stdout} stderr: ${result.stderr}');
+      return null;
     }
     return result.stdout.trim();
   }

--- a/packages/flutter_tools/lib/src/android/android_workflow.dart
+++ b/packages/flutter_tools/lib/src/android/android_workflow.dart
@@ -213,7 +213,7 @@ class AndroidWorkflow extends DoctorValidator implements Workflow {
       );
 
     final Version sdkManagerVersion = new Version.parse(androidSdk.sdkManagerVersion);
-    if (sdkManagerVersion.major < 26)
+    if (sdkManagerVersion == null || sdkManagerVersion.major < 26)
       // SDK manager is found, but needs to be updated.
       throwToolExit(
         'A newer version of the Android SDK is required. To update, run:\n'

--- a/packages/flutter_tools/test/android/android_sdk_test.dart
+++ b/packages/flutter_tools/test/android/android_sdk_test.dart
@@ -81,7 +81,7 @@ void main() {
       ProcessManager: () => processManager,
     });
 
-    testUsingContext('throws on sdkmanager version check failure', () {
+    testUsingContext('does not throw on sdkmanager version check failure', () {
       sdkDir = MockAndroidSdk.createSdkDirectory();
       Config.instance.setValue('android-sdk', sdkDir.path);
 
@@ -89,7 +89,7 @@ void main() {
       when(processManager.canRun(sdk.sdkManagerPath)).thenReturn(true);
       when(processManager.runSync(<String>[sdk.sdkManagerPath, '--version'], environment: argThat(isNotNull)))
           .thenReturn(new ProcessResult(1, 1, '26.1.1\n', 'Mystery error'));
-      expect(() => sdk.sdkManagerVersion, throwsToolExit(exitCode: 1));
+      expect(sdk.sdkManagerVersion, isNull);
     }, overrides: <Type, Generator>{
       FileSystem: () => fs,
       ProcessManager: () => processManager,

--- a/packages/flutter_tools/test/android/android_workflow_test.dart
+++ b/packages/flutter_tools/test/android/android_workflow_test.dart
@@ -132,7 +132,21 @@ void main() {
     when(sdk.sdkManagerPath).thenReturn('/foo/bar/sdkmanager');
     when(sdk.sdkManagerVersion).thenReturn('25.0.0');
 
-    expect(AndroidWorkflow.runLicenseManager(), throwsToolExit());
+    expect(AndroidWorkflow.runLicenseManager(), throwsToolExit(message: 'To update, run'));
+  }, overrides: <Type, Generator>{
+    AndroidSdk: () => sdk,
+    FileSystem: () => fs,
+    Platform: () => new FakePlatform()..environment = <String, String>{'HOME': '/home/me'},
+    ProcessManager: () => processManager,
+    Stdio: () => stdio,
+  });
+
+  testUsingContext('runLicenseManager errors correctly for null version', () async {
+    MockAndroidSdk.createSdkDirectory();
+    when(sdk.sdkManagerPath).thenReturn('/foo/bar/sdkmanager');
+    when(sdk.sdkManagerVersion).thenReturn(null);
+
+    expect(AndroidWorkflow.runLicenseManager(), throwsToolExit(message: 'To update, run'));
   }, overrides: <Type, Generator>{
     AndroidSdk: () => sdk,
     FileSystem: () => fs,

